### PR TITLE
Support imports for JS

### PR DESCRIPTION
**Pull request by PolinB**

### CODE REVIEW

File: indexation/build.gradle.kts [hunk 1]
```diff
@@ -9,16 +10,29 @@ plugins {
 
 dependencies {
     implementation(project(":common", configuration = "default"))
+    implementation("junit:junit:4.12")
+    implementation("org.jetbrains.intellij.deps:trove4j:1.0.20200330")
```
Comment (AlexanderPrendota) [inline indexation/build.gradle.kts:14]: why do u need it here?

File: indexation/src/main/kotlin/DescriptorsUtils.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    is ClassDescriptor -> {
+      if (visibility.isPublicAPI) {
+        ImportInfo(
+          importName = importName,
+          shortName = name.asString(),
+          fullName = name.asString(),
+          returnType = name.asString(),
+          icon = "class"
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/DescriptorsUtils.kt:49]: you have constants for this. But maybe it should be an enum class

File: indexation/src/main/kotlin/KotlinEnvironment.kt [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import org.jetbrains.kotlin.config.languageVersionSettings
+import org.jetbrains.kotlin.js.config.JSConfigurationKeys
+import org.jetbrains.kotlin.serialization.js.JsModuleDescriptor
+import org.jetbrains.kotlin.serialization.js.KotlinJavascriptSerializationUtil
+import org.jetbrains.kotlin.utils.KotlinJavascriptMetadataUtils
+import java.io.File
+
+class LibrariesFile(val jvm: File, val js: File)
+
+class KotlinEnvironmentConfiguration(val librariesFile: LibrariesFile) {
```
Comment (AlexanderPrendota) [inline indexation/src/main/kotlin/KotlinEnvironment.kt:23]: maybe in a common module?

File: src/main/kotlin/com/compiler/server/compiler/components/indexation/IndexationJsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
+package com.compiler.server.compiler.components.indexation
+
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+import javax.annotation.PostConstruct
+
+@Component
+class IndexationJsProvider(
```
Comment (AlexanderPrendota) [inline src/main/kotlin/com/compiler/server/compiler/components/indexation/IndexationJsProvider.kt:8]: it should be one component with two functions

**Review by AlexanderPrendota — CHANGES_REQUESTED** [review]

File: src/main/kotlin/com/compiler/server/compiler/components/IndexationProvider.kt [hunk 5]
```diff
@@ -1,38 +1,50 @@
...
 import org.springframework.beans.factory.annotation.Value
 import org.springframework.stereotype.Component
 import java.io.File
-import javax.annotation.PostConstruct
 
 @Component
 class IndexationProvider(
-  @Value("\${indexes.file}") private val indexesFileName: String
+  @Value("\${indexes.file}") private val indexesFileName: String,
+  @Value("\${indexesJs.file}") private val indexesJsFileName: String
```
Comment (bashor) [inline src/main/kotlin/com/compiler/server/compiler/components/IndexationProvider.kt:14]: JFTR: I think eventually we will need to add more platforms

File: src/test/kotlin/com/compiler/server/ImportTest.kt [hunk 6]
```diff
@@ -125,6 +125,114 @@ class ImportTest : BaseExecutorTest() {
...
+      completions = listOf(
+        "sin(x: Double)  (kotlin.math.sin)",
+        "sin(x: Float)  (kotlin.math.sin)"
+      ),
+      isJs = true
+    )
+  }
+
+  @Test
+  fun `open bracket after import completion js`() {
```
Comment (bashor) [inline src/test/kotlin/com/compiler/server/ImportTest.kt:197]: Would be nice to add a test for some API available only on K/JS.

Comment (bashor) [reply src/test/kotlin/com/compiler/server/ImportTest.kt:197]: https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.js/

Comment (PolinB) [reply src/test/kotlin/com/compiler/server/ImportTest.kt:197]: I tried to do tests for JS, but I couldn't find a specific function without arguments that I need to import

File: indexation/src/main/kotlin/DescriptorsUtils.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        ImportInfo(
+          importName = importName,
+          shortName = name.asString(),
+          fullName = name.asString(),
+          returnType = name.asString(),
+          icon = Icon.CLASS
+        )
+      } else null
+    }
+    else -> null
```
[the remark this answers is not in the document]
Comment (PolinB) [reply indexation/src/main/kotlin/DescriptorsUtils.kt:56]: I don't know what that means. Can you give me an example?

File: indexation/src/main/kotlin/KotlinEnvironment.kt [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package indexation
+
+import component.KotlinEnvironment
+import java.io.File
+
+class KotlinEnvironmentConfiguration(fileName: String) {
+  val kotlinEnvironment: KotlinEnvironment
+  init {
```
[the remark this answers is not in the document]
Comment (PolinB) [reply indexation/src/main/kotlin/KotlinEnvironment.kt]: Can you give me an example, please? I don't understand how do it.

File: indexation/src/main/kotlin/DescriptorsUtils.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (bashor) [reply indexation/src/main/kotlin/DescriptorsUtils.kt:56]: PropertyDescriptor

File: src/test/kotlin/com/compiler/server/ImportTest.kt (hunk 6 above)
Comment (bashor) [reply src/test/kotlin/com/compiler/server/ImportTest.kt:197]: Does the feature work only for functions?

Comment (bashor) [reply src/test/kotlin/com/compiler/server/ImportTest.kt:197]: You can try this one https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.js/json.html

File: common/src/main/kotlin/component/KotlinEnvironment.kt [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+     * See [org.jetbrains.kotlin.cli.common.arguments.CommonCompilerArguments] and
+     * [org.jetbrains.kotlin.cli.common.arguments.K2JVMCompilerArguments] for list of possible flags
+     */
+    private val additionalCompilerArguments: List<String> = listOf(
+      "-Xuse-experimental=kotlin.ExperimentalStdlibApi",
+      "-Xuse-experimental=kotlin.time.ExperimentalTime",
+      "-Xuse-experimental=kotlin.Experimental",
+      "-Xuse-experimental=kotlin.ExperimentalUnsignedTypes",
+      "-Xuse-experimental=kotlin.contracts.ExperimentalContracts",
+      "-Xuse-experimental=kotlin.experimental.ExperimentalTypeInference",
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply common/src/main/kotlin/component/KotlinEnvironment.kt:39]: @bashor we can just remove all experimental flags over here?

File: common/src/main/kotlin/component/KotlinEnvironment.kt [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+  private val configuration = createConfiguration()
+  val jsConfiguration = configuration.copy().apply {
+    put(CommonConfigurationKeys.MODULE_NAME, "moduleId")
+    put(JSConfigurationKeys.LIBRARIES, JS_LIBRARIES)
+  }
+
+  private val environment = KotlinCoreEnvironment.createForProduction(
+    parentDisposable = Disposer.newDisposable(),
+    configuration = configuration.copy(),
+    configFiles = EnvironmentConfigFiles.JVM_CONFIG_FILES
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply common/src/main/kotlin/component/KotlinEnvironment.kt:68]: is it important for the online version?

File: common/src/main/kotlin/component/KotlinEnvironment.kt [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+  )
+
+  private fun createConfiguration(): CompilerConfiguration {
+    val arguments = K2JVMCompilerArguments()
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply common/src/main/kotlin/component/KotlinEnvironment.kt:72]: is it important for the online version?

File: indexation/src/main/kotlin/KotlinEnvironment.kt [hunk 12]
```diff
@@ -0,0 +1,21 @@
+package indexation
+
+import component.KotlinEnvironment
+import java.io.File
+
+class KotlinEnvironmentConfiguration(fileName: String) {
```
[the remark this answers is not in the document]
Comment (AlexanderPrendota) [reply indexation/src/main/kotlin/KotlinEnvironment.kt]: yes

**Comment by AlexanderPrendota** [comment]
@PolinB tests =)

**Review by AlexanderPrendota — APPROVED** [review]

**Pull request merged**